tensornetwork/tn_keras/mpo.py

```diff
@@ -0,0 +1,167 @@
+import tensornetwork as tn
+from tensornetwork.network_components import Node
+import numpy as np
+import math
+
+
+class DenseMPO(Layer):
+  """Matrix Product Operator (MPO) TN layer.
+
+  Example:
+
+  ```python
+  # as first layer in a sequential model:
+  model = Sequential()
+  model.add(
+    DenseMPO(1024, num_nodes=4, bond_dim=8, activation='relu', input_dim=1024))
+  # now the model will take as input arrays of shape (*, 1024)
+  # and output arrays of shape (*, 1024)
+  # after the first layer, you don't need to specify
+  # the size of the input anymore:
+  model.add(DenseMPO(1024, num_nodes=4, bond_dim=8, activation='relu'))
+  ```
+
+  Args:
+    output_dim: Positive integer, dimensionality of the output space.
+    num_nodes: Positive integer, number of nodes in the MPO.
+      Note input_dim**(1. / num_nodes) and output_dim**(1. / num_nodes)
+      must both be round.
+    bond_dim: Positive integer, size of the intermediate dimension.
+    activation: Activation function to use.
+      If you don't specify anything, no activation is applied
+      (ie. "linear" activation: `a(x) = x`).
+    use_bias: Boolean, whether the layer uses a bias vector.
+    kernel_initializer: Initializer for the node weight matrices.
+    bias_initializer: Initializer for the bias vector.
+
+  Input shape:
+    2D tensor with shape: `(batch_size, input_dim)`.
+
+  Output shape:
+    2D tensor with shape: `(batch_size, output_dim)`.
+  """
+
+  def __init__(self,
+               output_dim: int,
+               num_nodes: int,
+               bond_dim: int,
+               use_bias: Optional[bool] = True,
+               activation: Optional[Text] = None,
+               kernel_initializer: Optional[Text] = 'glorot_uniform',
+               bias_initializer: Optional[Text] = 'zeros',
+               **kwargs) -> None:
+    if 'input_shape' not in kwargs and 'input_dim' in kwargs:
+      kwargs['input_shape'] = (kwargs.pop('input_dim'),)
+
+    assert num_nodes > 2, 'Need at least 2 nodes to create MPO'
+
+    self.output_dim = output_dim
+    self.num_nodes = num_nodes
+    self.bond_dim = bond_dim
+    self.nodes = []
+
+    self.use_bias = use_bias
+    self.activation = activations.get(activation)
+    self.kernel_initializer = initializers.get(kernel_initializer)
+    self.bias_initializer = initializers.get(bias_initializer)
+    super(DenseMPO, self).__init__(**kwargs)
+
+  def build(self, input_shape: List[int]) -> None:
+    # Disable the attribute-defined-outside-init violations in this function
+    # pylint: disable=attribute-defined-outside-init
+    if input_shape[-1] is None:
+      raise ValueError('The last dimension of the inputs to `Dense` '
+                       'should be defined. Found `None`.')
+
+    def is_perfect_root(n, n_nodes):
+      root = n**(1. / n_nodes)
+      return round(root)**n_nodes == n
+
+    # Ensure the MPO dimensions will work
+    assert is_perfect_root(input_shape[-1],
```

Comment:
Nit: It's generally good to print the values in the error message - if a user gets this, it's likely they will have to tear apart their code or add print statements to see what the values actually were.

Reply:
Good call, updated the error message